### PR TITLE
Admin spell books

### PR DESCRIPTION
**Pull request by opsaaaaa**
A command utility used to create command books (spell books)
Its helpful for admins on survival worlds.
I use it to spawn and kill player bot in standard locations at farms, but the its not limited to that usecase.
```
/spellbook <book> set <title> <"command">
/spellbook farms set "spawn slime farm bot" "/execute in overworld run player SlimeBot spawn at -50.50 82 24"
/spellbook farms set "kill slime farm bot" "/player SlimeBot kill"
/spellbook farms give
```
This will generate a book with clickable text that runs the above commands.
Spell books placed in lecterns will automatically update to new versions of the spell book when opened.

### CODE REVIEW

File: programs/survival/spellbook.sc
Comment (opsaaaaa) [inline programs/survival/spellbook.sc:185]: This line is a little gross in my opinion. each book page is stored as a string but can actually have more nbt structure within it. I am treating the whole nbt tag as a string until its used. 

the other way would be to convert it to nbt and then convert each page back to a string when your done. that would involve some rewriting bits.

**Comment by Firigion** [comment]
Looks cool, I'll test it when I get some time. For now, how about adding a hover message to the command? You could do that either in the same command (`/spellbook <book> set <title> <"command"> [<hover_text>]`) or in a separate one (`/spellbook <book> set <title> <"hover">`, will fail if `title` has not been created yet). I'd suggest the default hover test is just the command, so people cna see stuff like what coords the bot is spawned at and such.

**Comment by opsaaaaa** [comment]
For anyone who has already done some testing with this script, run this command to migrate the book files from the app root to app/books 
```
/script in spellbook invoke migrate_books_folder_location()
```

**Comment by opsaaaaa** [comment]
So turns out there is still some breaking changes I want to make. I intend to use a migrate function like I did earlier but I think it would be cleaner to hold off on merging/ reviewing this until I make those changes. 

Also I moved the app from survival into utilities. I think the readme descriptions for the survival apps are very helpful. I would not mind including a description for this app in the utilities readme but that does not appear to be standard practice.

**Comment by opsaaaaa** [comment]
ok I am happy with where this app is at now and I am done making changes.

**Comment by opsaaaaa** [comment]
If you have existing spellbooks and want to test the updated version you can run `/script in spellbook invoke migrate_spells_data_structure` to migrate your old spells up to this version

File: programs/survival/README.md
Comment (Firigion) [inline programs/survival/README.md:299]: Missing input in the readme?

Comment (opsaaaaa) [reply programs/survival/README.md:299]: I moved the app from `survival/` into `utilities/`, the app feels more like a utility then a survival feature. so I decided to remove the survival readme entire I created. Apparently I left a new line. 
I was thought about adding it to the utilities readme entire, however that file is currently empty. so I was uncertain how welcome setting that precedent would be. In hindsight i should have asked about it on the discord server.

Would it be cool if I added the readme entire into the utilities readme file and formatted it like the survival one?

Comment (altrisi) [reply programs/survival/README.md:299]: There is a readme being added to utilities in #289, please don't add one here or they will conflict.

If you want though you can leave the description here and it can be added later.

Comment (opsaaaaa) [reply programs/survival/README.md:299]: I assume 'here' means here in the `survival/README.md`

Comment (altrisi) [reply programs/survival/README.md:299]: I mean in this PR, that PR linked adds a readme to utilities.

Comment (opsaaaaa) [reply programs/survival/README.md]: I added a description to the survival readme as well as in the head of the spellbook script

**Comment by opsaaaaa** [comment]
This is the spellbook description to be added to the utilities readme when its ready. 

### [spellbook.sc](https://github.com/gnembon/scarpet/blob/master/programs/utilities/spellbook.sc):
#### By Opsaaaaa
	A in game utility for creating, using, and updating command books (spellbooks)
	You create a new spell in a book with this command
	`/spellbook <book> set <spell> </command>` 
	`/spellbook cat_spells set "spawn cat" /summon cat`
	You can continue adding spells to existing spellbooks. 
	`/spellbook cat_spells set "spawn Tabby" /summon cat ~ ~ ~ {CatType:2}`
	Then you can give yourself a copy of the spellbook with the spellbook give command,
	`/spellbook cat_spells give`
	Any changes you make to your spells will automatically update your spellbook when opened.
	You can find more info in-game with the help command',
	`/spellbook help <page>' Pages: `(main), basics, shorthands, customize, commands`

File: programs/utilities/spellbook.sc
Comment (Firigion) [inline programs/utilities/spellbook.sc:67]: make these two be hyperkins? Click the name and have the help page in question showed

File: programs/utilities/spellbook.sc
Comment (Firigion) [inline programs/utilities/spellbook.sc:327]: Maybe add a confirmation message to the delete action. This is very deffinitive to be able to do it by mistake.

**Review by Firigion — APPROVED** [review]
It's a big app, I didn't get to test it too thoroughly. I think the best move is to push it to the repo and have eventual bug reposrts from users trigger new PRs to fix them. I didn't find anything critical. UI is suboptimal, but alos very hard to do better, so I'm not conna complain.

File: programs/utilities/spellbook.sc
Comment (opsaaaaa) [reply programs/utilities/spellbook.sc:327]: Yeah that's a good idea

File: programs/utilities/spellbook.sc
Comment (opsaaaaa) [reply programs/utilities/spellbook.sc:128]: That would be better, Ill do that this weekend.

Comment (opsaaaaa) [reply programs/utilities/spellbook.sc:128]: Mostly that's the fist time I was really using format and I wasn't that familiar with it, then after I prioritized other feature.

Comment (opsaaaaa) [reply programs/utilities/spellbook.sc]: Ok... I tired that and I didn't like it. The differences between examples and references made it feel gross.
`/spellbook <book> set <spell> </command>` vs `/spellbook cat_spells set "spawn cat" /summon cat`
Return characters `\n` also posed a problem.
I did standardize them with a function. It might be slightly easier to add in the future now.

tldr: Its good enough as is. After trying it I decided not to add clipboard or suggestion support to the help pages.

**Review by altrisi — APPROVED** [review]
This app is amazing, now I feel really bad for not checking it earlier.
It looks a bit complex at first, but gets simple fast enough.

One thing I've found is that idk if it's just I'm doing it wrong but I felt like some actions were being added to all books or something like that, but other than that it's amazing.

Also small suggestion that isn't really needed but would be nice, what would you think of the command ref suggesting the command on click?

**Comment by opsaaaaa** [comment]
@altrisi 
Thank, yeah I thought it was handy.

> what would you think of the command ref suggesting the command on click?

I'm honestly not quite totally sure what you mean. Tho it's been awhile since I touched the app.

> I felt like some actions were being added to all books 

So there is a feature that syncs spellbook with the same name. Perhaps I failed to make that clear anywhere. `/spellbook farms bot ...` adds the new spell to the global spellbook titles farms. 
All of the other farms spell books in the world, when opened will automatically update with the new spell. So to keep spell sets separate you can use different titles. 'farms' 'warps' and so on.

The use case at the time, I hade a bunch of players on the server who liked to build afk farms. And so I wanted to give all the players a way to easily turn off other people farms when causing lag. And then have the farm owner not get annoyed about traveling to the farm any time he wanted to turn it back on. So everyone had the same synced farms spellbook.

However if you mean that some actions applied to both the "farms" and "warps" spellbooks, that would be somewhat odd. but if you saw that it could certainly be a bug.

**Comment by altrisi** [comment]
Quite sure it was after adding to differently named books (I noticed and liked a lot the updating feature). I'll try to make some better repro steps later, maybe I did something wrong.

With the suggestions I mean that when you click a command in the help menu for example, it writes it into your chat box so you can start using it already, just a small thing I thought would be nice at the start.

**Comment by opsaaaaa** [comment]
Aww yes that would be nice. I think that can be an addition after the app is merged tho.

**Comment by altrisi** [comment]
Idk why this wasn't merged already having the 2 reviews

**Pull request merged**
